fabric-dimensions-v1/src/main/java/net/fabricmc/fabric/impl/dimension/FabricDimensionInternals.java

```diff
@@ -0,0 +1,96 @@
+     *   and use it to call the entity placers
+     * This lets us use the exact same logic for any entity and prevent the vanilla getPortal (which has unwanted
+     * side effects) from running, while keeping the patches minimally invasive.
+     *
+     * Shortcomings: bugs may arise if another patch cancels the teleportation method between
+     * #prepareDimensionalTeleportation and #tryFindPlacement, AND a mod calls PortalForcer#getPortal directly
+     * right after.
+     */
+
+    public static void prepareDimensionalTeleportation(Entity entity) {
+        Preconditions.checkNotNull(entity);
+        PORTAL_ENTITY.set(entity);
+        // Set values used by `PortalForcer#changeDimension` to prevent a NPE crash.
+        EntityHooks access = ((EntityHooks) entity);
+        if (entity.getLastPortalDirectionVector() == null) {
+            access.setLastPortalDirectionVector(entity.getRotationVector());
+        }
+        if (entity.getLastPortalDirection() == null) {
+            access.setLastPortalDirection(entity.getHorizontalFacing());
+        }
+    }
+
+    /* Nullable */
+    public static BlockPattern.TeleportTarget tryFindPlacement(ServerWorld destination, Direction portalDir, double portalX, double portalY) {
+        Preconditions.checkNotNull(destination);
+        Entity teleported = PORTAL_ENTITY.get();
+        PORTAL_ENTITY.set(null);
+        // If the entity is null, the call does not come from a vanilla context
+        if (teleported == null) {
+            return null;
+        }
+
+        // Custom placement logic, falls back to default dimension placement if no placement or target found
+        EntityPlacer customPlacement = CUSTOM_PLACEMENT.get();
+        if (customPlacement != null) {
+            BlockPattern.TeleportTarget customTarget = customPlacement.placeEntity(teleported, destination, portalDir, portalX, portalY);
+            if (customTarget != null) {
+                return customTarget;
+            }
+        }
+
+        // Default placement logic, falls back to vanilla if not a fabric dimension
+        DimensionType dimType = destination.getDimension().getType();
+        if (dimType instanceof FabricDimensionType) {
+            BlockPattern.TeleportTarget defaultTarget = ((FabricDimensionType) dimType).getDefaultPlacement().placeEntity(teleported, destination, portalDir, portalX, portalY);
+            if (defaultTarget == null) {
+                throw new IllegalStateException("Mod dimension " + DimensionType.getId(dimType) + " returned an invalid teleport target");
+            }
+            return defaultTarget;
+        }
+
+        // Vanilla / other implementations logic, undefined behaviour on custom dimensions
+        return null;
+    }
+
+    @SuppressWarnings("unchecked")
+    public static <E extends Entity> E changeDimension(E teleported, DimensionType dimension, EntityPlacer placement) {
```

Comment:
Validate it's being executed for a server-side entity? Assert being on the server thread?